ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,420 @@
+    }
+}
+
+impl AllocatorSelector for Untracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_UNTRACKED_ALLOCATOR
+    }
+}
+
+/// Kernel Virtual Area.
+///
+/// A tracked kernel virtual area (`Kva<Tracked>`) manages a range of memory in
+/// `TRACKED_MAPPED_PAGES_RANGE`. It can map a inner part or all of its virtual memory
+/// to some physical tracked pages.
+///
+/// A untracked kernel virtual area (`Kva<Untracked>`) manages a range of memory in
+/// `VMALLOC_VADDR_RANGE`. It can map a inner part or all of its virtual memory to
+/// some physical untracked pages.
+#[derive(Debug)]
+pub struct Kva<M: AllocatorSelector + 'static> {
+    range: Range<Vaddr>,
+    phantom: PhantomData<M>,
+}
+
+impl<M: AllocatorSelector + 'static> Kva<M> {
+    pub fn new(size: usize) -> Self {
+        let allocator = M::select_allocator();
+        let range = allocator.alloc(size).unwrap();
+        Self {
+            range,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn start(&self) -> Vaddr {
+        self.range.start
+    }
+
+    pub fn end(&self) -> Vaddr {
+        self.range.end
+    }
+
+    pub fn range(&self) -> Range<Vaddr> {
+        self.range.start..self.range.end
+    }
+
+    fn query_page(&self, addr: Vaddr) -> PageTableItem {
+        assert!(self.start() <= addr && self.end() >= addr);
+        let start = addr.align_down(PAGE_SIZE);
+        let vaddr = start..start + PAGE_SIZE;
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let mut cursor = page_table.cursor(&vaddr).unwrap();
+        cursor.query().unwrap()
+    }
+}
+
+impl Kva<Tracked> {
+    /// Maps pages into the kernel virtual area.
+    pub fn map_pages<T: PageMeta>(
+        &mut self,
+        range: Range<Vaddr>,
+        pages: impl Iterator<Item = Page<T>>,
+        prop: PageProperty,
+    ) {
+        assert!(self.start() <= range.start && self.end() >= range.end);
+        let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+        let mut cursor = page_table.cursor_mut(&range).unwrap();
+        let flusher = TlbFlusher::new(CpuSet::new_full(), disable_preempt());
+        let mut va = self.start();
+        for page in pages.into_iter() {
+            // SAFETY: The constructor of the `Kva<Tracked>` structure has already ensured this
+            // mapping does not affect kernel's memory safety.
+            if let Some(old) = unsafe { cursor.map(page.into(), prop) } {
+                flusher.issue_tlb_flush_with(TlbFlushOp::Address(va), old);
+                flusher.dispatch_tlb_flush();
+            }
+            va += PAGE_SIZE;
+        }
+        flusher.issue_tlb_flush(TlbFlushOp::Range(range));
+        flusher.dispatch_tlb_flush();
+    }
+
+    /// Gets the mapped tracked page.
+    ///
+    /// This function returns None in the following cases:
+    ///  - If the address is not mapped (`NotMapped`).
+    ///  - If the address is mapped to a `MappedUntracked` page.
```

Comment:
`Kva<Tracked>` does not allow mapping untracked pages in the first place. So don't return `None` when you discover that it's untracked. Panic when it's not tracked.